File: Employee Absenteeism.R
rm(list=ls(all=T))

getwd()

#Load Libraries
x = c("ggplot2", "corrgram", "DMwR", "caret", "randomForest", "unbalanced", "C50", "dummies", "e1071", "Information",
      "MASS", "rpart", "gbm", "ROSE", 'sampling', 'DataCombine', 'inTrees','readxl')

#install.packages('readxl')
lapply(x, require, character.only = TRUE)
rm(x)

## Read the Data
df = read.csv("Absent_to_work.csv", header = T, na.strings = c(" ", "", "NA"))

#Check number of unique values in each column
for(i in 1:ncol(df))
{
  print(colnames(df[i]))
  print(length(unique(df[,i])))
}


######## Missing Values #############
missing_val = data.frame(apply(df,2,function(x){sum(is.na(x))}))
missing_val$Columns = row.names(missing_val)
names(missing_val)[1] =  "Missing_percentage"
missing_val$Missing_percentage = (missing_val$Missing_percentage/nrow(df)) * 100
missing_val = missing_val[order(-missing_val$Missing_percentage),]
row.names(missing_val) = NULL
missing_val = missing_val[,c(2,1)]

#Convert it to dataframe
df = as.data.frame(df)


# kNN Imputation
df = knnImputation(df, k=3)



######### Outlier Analysis ######################

cnames = c('Transportation.expense', 'Distance.from.Residence.to.Work','Service.time', 'Age', 'Work.load.Average.day'
            ,'Hit.target','Weight','Height','Body.mass.index','Absenteeism.time.in.hours')



for (i in 1:length(cnames))
{
  assign(paste0("gn",i), ggplot(aes_string(y = (cnames[i]), x = "Absenteeism.time.in.hours"), data = subset(df))+
           stat_boxplot(geom = "errorbar", width = 0.5) +
           geom_boxplot(outlier.colour="red", fill = "grey" ,outlier.shape=18,
                        outlier.size=1, notch=FALSE) +
           theme(legend.position="bottom")+
           labs(y=cnames[i],x="Absenteeism.time.in.hours")+
           ggtitle(paste("Box plot for",cnames[i])))
}


## Plotting plots together
gridExtra::grid.arrange(gn1,gn2,ncol=2)
gridExtra::grid.arrange(gn3,gn4,ncol=2)
gridExtra::grid.arrange(gn5,gn6,ncol=2)
gridExtra::grid.arrange(gn7,gn8,ncol=2)
gridExtra::grid.arrange(gn9,gn10,ncol=2)



for(i in cnames)
{
  val = df[,i][df[,i] %in% boxplot.stats(df[,i])$out]
  df[,i][df[,i] %in% val] = NA
  
}
df = knnImputation(df, k = 3)



############## Feature Selection ##########################


corrgram(df[,cnames], order = F, upper.panel = panel.pie, text.panel = panel.txt, main = "Correlation Plot")

##Dimensionality Reduction
df = subset(df, select = -c(Body.mass.index ))


############## Feature Scaling ###############################

for(i in colnames(df))
{
  print(i)
  df[,i] = (df[,i] - min(df[,i]))/ (max(df[,i] - min(df[,i])))
}


############# Model Development ##################################

train_index = sample(1:nrow(df), 0.8 * nrow(df))
train = df[train_index,]
test = df[-train_index,]


#### r part regression - Decision Tree
fit_DT = rpart(Absenteeism.time.in.hours ~., data = train, method = "anova")

#predict new case
predictions = predict(fit, test[,-20])
summary(fit_DT)


rmse = function(error)
{
  sqrt(mean(error^2))
}

mae = function(error)
{
  mean(abs(error))
}


error = test$Absenteeism.time.in.hours - predictions
rmse(error)
mae(error)

### Decision Tree 

#RMSE = 0.171
#MAE = 0.118




#### Random Forest

#Develop Model on training data
fit_RF = randomForest(Absenteeism.time.in.hours~., data = train)

#Lets predict for training data
pred_RF= predict(fit_RF, test[,-20])

error = test$Absenteeism.time.in.hours - pred_RF
rmse(error)
mae(error)


### Random Forest 

#RMSE = 0.162
#MAE = 0.115


#### Linear Regression

#Develop Model on training data
fit_LR = lm(Absenteeism.time.in.hours ~ ., data = train)

#Lets predict for training data
pred_LR = predict(fit_LR, test[,-20])



error = test$Absenteeism.time.in.hours - pred_LR
rmse(error)
mae(error)


### Linear Regression 

#RMSE = 0.175
#MAE = 0.127

File: _Employee Absenteeism.py

#Import required libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from fancyimpute import KNN
from scipy.stats import chi2_contingency


# Load the dataset


df = pd.read_excel('Absenteeism_at_work_Project.xls')



#  Intial Analysis

df.info()

df.shape


df.describe()




############### Missing Value Analysis ######################



missing_val = pd.DataFrame(df.isnull().sum())

missing_val = missing_val.reset_index()

missing_val = missing_val.rename(columns = {'index': 'Variables', 0:'Missing_Percentage'})

missing_val['Missing_Percentage'] = (missing_val['Missing_Percentage']/ len(df))*100

missing_val =  missing_val.sort_values('Missing_Percentage', ascending=False).reset_index(drop = True)


#Impute Method
#df['Absenteeism time in hours'][28]
#Actual Value = 8
#mean = 6.97
#median = 3.0
#KNN = 7.11



#Knn_imputation for missing values
df = pd.DataFrame(KNN(k=3).complete(df), columns=df.columns)



#Round of the numeric values

for i in df.columns:
    df[i] = df[i].round()


########### Exploratory Data Analysis ############


cnames = ['Transportation expense', 'Distance from Residence to Work','Service time', 'Age', 'Work load Average/day '
          ,'Hit target','Weight','Height','Body mass index','Absenteeism time in hours']


#Distribution plot with kernel density for all the predictors
for i in cnames:
    k = sns.distplot(df[i])
    plt.show()
  


#Graphs to analyse the data

for i in df.columns:
    sns.barplot(x = i, y = "Absenteeism time in hours", data = df)
    plt.tight_layout()
    plt.show()



for i in df.columns:
    plt.figure(figsize = (15,6))
    sns.barplot(x = i, y = "Absenteeism time in hours", data = df)
    plt.tight_layout()



plt.figure(figsize = (18,6))
sns.barplot(x='Age', y='Absenteeism time in hours', hue = 'Service time',data=df)
plt.tight_layout()



################ Outlier Analysis #########################


for i in cnames:
    plt.boxplot(df[i])
    plt.xlabel(i)
    plt.show()
  


# Continious variables


cnames = ['Transportation expense', 'Distance from Residence to Work','Service time', 'Age', 'Work load Average/day '
          ,'Hit target','Weight','Height','Body mass index','Absenteeism time in hours']



#With outliers
for i in cnames:
    q75, q25 = np.percentile(df.loc[:,i], [75 ,25])
    iqr = q75 - q25

    minimum = q25 - (iqr*1.5)
    maximum = q75 + (iqr*1.5)
    
    print(minimum)
    print(maximum)
   
    print(len(df[df.loc[:,i] < minimum]))
    print(len(df[df.loc[:,i] > maximum]))
    
    df[i][df.loc[:,i] < minimum] = np.nan
    df[i][df.loc[:,i] > maximum] = np.nan

#Impute NA with KNN_Imputation
df = pd.DataFrame(KNN(k = 3).complete(df), columns = df.columns)


############# Feature Selection ###########


df_corr = df.loc[:,cnames]

corr = df_corr.corr()

#Plot using seaborn library
sns.heatmap(corr, mask=np.zeros_like(corr, dtype=np.bool), cmap=sns.diverging_palette(220, 10, as_cmap=True),
            square=True)
#plt.savefig("heatmap.png")



catnames = ['ID','Reason for absence', 'Month of absence','Age', 'Day of the week','Seasons','Disciplinary failure', 'Education', 'Son', 'Social drinker',
       'Social smoker', 'Pet',]



for i in catnames:
    print(i)
    chi2, p, dof, ex = chi2_contingency(pd.crosstab(df['Absenteeism time in hours'], df[i]))
    print(p)



# Dimensionality Reduction
df = df.drop(['Body mass index'], axis = 1)



df.shape



################ Feature Scaling #################


for i in df.columns:
    print(i)
    df[i] = (df[i] - np.min(df[i])) / (np.max(df[i]) - np.min(df[i]))




################## Model Development ##############



from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.cross_validation import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error



#split the dataset into train and test dataset 
X = df.iloc[:,0:19]
y = df.iloc[:,19]


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33)


####Decision Tree Regressor


DT_reg = DecisionTreeRegressor(max_depth=2)

DT_reg.fit(X_train,y_train)

#predict test cases
Prediction_DT = DT_reg.predict(X_test)


# Calculate RMSE and MAE

rmse_for_test = np.sqrt(mean_squared_error(y_test,Prediction_DT))

MAE = mean_absolute_error(y_test, Prediction_DT)


##Error metrics for Decision Tree
#MAE =  0.13
#RMSE = 0.18



#### Random Forest


RF_Reg = RandomForestRegressor()

RF_Reg.fit(X_train, y_train)

#predict the test cases
Prediction_RF = RF_Reg.predict(X_test)


rmse_for_test_RF = np.sqrt(mean_squared_error(y_test,Prediction_RF))


MAE = mean_absolute_error(y_test, Prediction_RF)



##Error metrics for Random Forest
#MAE =  0.12
#RMSE = 0.17


#### Linear Regression


import statsmodels.api as sm

model = sm.OLS(y_train, X_train).fit()

model.summary()

Prediction_LR = model.predict(X_test)


rmse_for_test_LR = np.sqrt(mean_squared_error(y_test,Prediction_LR))


MAE = mean_absolute_error(y_test, Prediction_LR)



##Error metrics for Random Forest
#MAE =  0.14
#RMSE = 0.18



